File: final/pb16.cpp
#include <iostream>
#include <vector>
using namespace std;


using VE = vector<int>;


int n, d;
VE V;
VE usat;


void f(int i) {
  if (i == n) {
    for (int x : V) cout << char('A' + x);
    cout << endl;
    return;
  }

  for (int x = 0; x < n; ++x)
    if ((x <= i - d or x >= i + d) and not usat[x]) {
      usat[x] = true;
      V[i] = x;
      f(i + 1);
      usat[x] = false;
    }
}


int main() {
  cin >> n >> d;
  V = VE(n);
  usat = VE(n, false);
  f(0);
}

File: final/pb16.py
def rec(res, usat, length):
    if length == n:
        print("".join(res))
        return

    for i in range(n):
        if not usat[i] and abs(i - length) >= d:
            res[length] = chr(ord('A') + i)
            usat[i] = True
            rec(res, usat, length + 1)
            usat[i] = False


n = int(input())
d = int(input())

res = ["" for i in range(n)]
usat = [False for i in range(n)]

rec(res, usat, 0)
